torch/csrc/autograd/profiler_kineto.h

```diff
@@ -382,6 +382,9 @@ TORCH_API void enableProfilerWithEventPostProcess(
 
 TORCH_API std::unique_ptr<ProfilerResult> disableProfiler();
 
+TORCH_API void registerProfilerFactory(
+    torch::profiler::impl::kineto::child_activity_profiler_factory_t factory);
```

Comment:
`profiler_kineto.h` cannot directly include `kineto_shim.h` anymore. The reason is that we have started to make limited use of kineto headers in our internal implementation (just to get access to the enums) but that requires a strict separation of the implementation headers (like `kineto_shim.h`) and the API headers (like `profiler_kineto.h`). Otherwise kineto symbols leak out, things fail to link, and it's a mess. I apologize for the inconvenience; this changed since I first recommended this approach.

That said, I don't think that this is insurmountable. (The issue above was more about not being able to find function definitions.) If this is a one off we can probably just add some forward declares in this file and pass the factory through by pointer. On the other hand, if you're expecting to need more interaction with Kineto through pytorch we'll probably need to split `kineto_shim.h` in a "forward declares + pointers" header that is safe for the public API and a private only "actually include some kineto headers" header. WDYT?

Reply:
Hi @robieta that's ok in this scenario, because only one symbol is required. Can you please take a look at the latest commit: only forward-declaring IActivityProfiler and creating separate factory function, no kineto_shim dependencies. Is this compatible with current code organization model?

Reply:
It's failing with:
```
error: could not convert 'factory' from 'torch::autograd::profiler::child_activity_profiler_factory_t {aka std::function<std::unique_ptr<libkineto::IActivityProfiler>()>}' to 'torch::profiler::impl::kineto::child_activity_profiler_factory_t {aka torch::profiler::impl::kineto::DummyChildActivityProfilerFactory}'
   torch::profiler::impl::kineto::registerProfilerFactory(factory);
```
Let me quickly see how painful the header split is. (I suspect not very.)